File: swift/hamming/hamming.swift
//
//  hamming.swift
//  helloWorld
//
//  Created by Raul Galindo on 29/01/16.
//  Copyright © 2016 exercism.io. All rights reserved.
//

import Foundations

class Hamming {

    static func compute(FirstDnaStrand: String, against SecondDnaStrand: String) -> Int? {

        let firstDnaStrandArray = Array(FirstDnaStrand.characters)
        let secondDnaStrandArray = Array(SecondDnaStrand.characters)

        if firstDnaStrandArray.count != secondDnaStrandArray.count {
            return nil
        }

        var mutationsCount = 0

        for (index, nucleotide) in firstDnaStrandArray.enumerate() {
            let againstNucleotide = secondDnaStrandArray[index]
            if (nucleotide != againstNucleotide) {
             mutationsCount += 1
            }
        }

        return mutationsCount
    }

}

File: ruby/hamming/hamming.rb
class Hamming

  attr_accessor :strand1, :strand2

  def self.compute strand1, strand2
    calculator = new strand1, strand2

    calculator.perform
  end

  def initialize strand1, strand2
    @strand1, @strand2  = strand1, strand2
  end

  def perform
    (0...strand_size).count do |index|
      strand1.at( index ) != strand2.at( index )
    end
  end

private

  def strand_size
    [ strand1.length, strand2.length ].min
  end

end